service/worker/deployment/deployment_util.go

```diff
@@ -0,0 +1,226 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	updatepb "go.temporal.io/api/update/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	deployspb "go.temporal.io/server/api/deployment/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/payload"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/sdk"
+	"go.temporal.io/server/common/searchattribute"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type DeploymentClient interface {
+	RegisterWorker(
+		ctx context.Context,
+		taskQueueName string,
+		taskQueueType enumspb.TaskQueueType,
+		pollTimestamp time.Time,
+	) error
+}
+
+// implements DeploymentClient
+type DeploymentWorkflowClient struct {
+	namespaceEntry *namespace.Namespace
+	deployment     *commonpb.WorkerDeployment
+	historyClient  resource.HistoryClient
+}
+
+func NewDeploymentWorkflowClient(
+	namespaceEntry *namespace.Namespace,
+	deployment *commonpb.WorkerDeployment,
+	historyClient resource.HistoryClient,
+) *DeploymentWorkflowClient {
+	return &DeploymentWorkflowClient{
+		namespaceEntry: namespaceEntry,
+		deployment:     deployment,
+		historyClient:  historyClient,
+	}
+}
+
+func (d *DeploymentWorkflowClient) RegisterTaskQueueWorker(
+	ctx context.Context,
+	taskQueueName string,
+	taskQueueType enumspb.TaskQueueType,
+	pollTimestamp *timestamppb.Timestamp,
+	maxIDLengthLimit int,
+) error {
+	sa := &commonpb.SearchAttributes{}
+	searchattribute.AddSearchAttribute(&sa, searchattribute.TemporalNamespaceDivision, payload.EncodeString(DeploymentNamespaceDivision))
```

Comment:
I would move this down to line 113, right before constructing startReq, just to make the code flow better